File: rust/src/lru.rs
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

#[derive(Clone)]
pub struct Node<T> {
    pub value: T,
    pub next: Option<Rc<RefCell<Node<T>>>>,
    pub prev: Option<Rc<RefCell<Node<T>>>>,
}

fn create_node<V>(value: V) -> Node<V> {
    Node {
        value,
        next: None,
        prev: None,
    }
}

pub struct LRU<K, V>
where
    K: std::cmp::Eq + std::hash::Hash + Clone,
    V: Clone,
{
    capacity: usize,
    length: usize,
    head: Option<Rc<RefCell<Node<V>>>>,
    tail: Option<Rc<RefCell<Node<V>>>>,
    lookup: HashMap<K, Rc<RefCell<Node<V>>>>,
    reverse_lookup: HashMap<Rc<RefCell<Node<V>>>, K>,
}

impl<K, V> LRU<K, V>
where
    K: std::cmp::Eq + std::hash::Hash + Clone,
    V: Clone,
{
    pub fn new(capacity: usize) -> LRU<K, V> {
        LRU {
            capacity,
            length: 0,
            head: None,
            tail: None,
            lookup: std::collections::HashMap::new(),
            reverse_lookup: std::collections::HashMap::new(),
        }
    }

    pub fn update(&mut self, key: K, value: V) {
        let mut node = self.lookup.get(&key).cloned();

        if node.is_none() {
            let new_node = Box::new(create_node(value));
            self.length += 1;
            self.prepend(new_node.clone());
            self.trim_cache();
            self.lookup.insert(key.clone(), new_node.clone());
            self.reverse_lookup.insert(new_node, key);
        } else {
            let mut node = node.unwrap();
            self.detach(&mut node);
            self.prepend(node.clone());
            node.value = value;
        }
    }

    pub fn get(&mut self, key: K) -> Option<&V> {
        let node = self.lookup.get(&key);

        match node {
            None => None,
            Some(node) => {
                let mut node = node.clone();
                self.detach(&mut node);
                self.prepend(node.clone());
                Some(&node.value)
            }
        }
    }

    fn detach(&mut self, node: &mut Box<Node<V>>) {
        if let Some(prev) = node.prev.as_mut() {
            prev.next = node.next.clone();
        }
        if let Some(next) = node.next.as_mut() {
            next.prev = node.prev.clone();
        }

        if self.length == 1 {
            self.head = None;
            self.tail = None;
        }

        if self.head.as_ref().map_or(false, |head| head.as_ref() as *const _ == node.as_ref() as *const _) {
            self.head = node.next.clone();
        }

        if self.tail.as_ref().map_or(false, |tail| tail.as_ref() as *const _ == node.as_ref() as *const _) {
            self.tail = node.prev.clone();
        }

        node.next = None;
        node.prev = None;
    }

    fn prepend(&mut self, node: Box<Node<V>>) {
        match self.head {
            None => {
                self.head = Some(node.clone());
                self.tail = Some(node);
            }
            Some(_) => {
                node.next = self.head.clone();
                self.head.as_mut().unwrap().prev = Some(node.clone());
                self.head = Some(node);
            }
        }
    }

    fn trim_cache(&mut self) {
        if self.length <= self.capacity {
            return;
        }

        let tail = self.tail.take().unwrap();
        self.detach(&mut tail.clone());

        if let Some(key) = self.reverse_lookup.get(&tail) {
            let key = key.clone();
            self.lookup.remove(&key);
            self.reverse_lookup.remove(&tail);
        }
        self.length -= 1;
    }
}

File: java/src/LRU.java
package src;
import java.util.HashMap;
import java.util.Map;

class Node<T> {
    T value;
    Node<T> next;
    Node<T> prev;
}

public class LRU<K, V> {
    private int length;
    private Node<V> head;
    private Node<V> tail;

    private Map<K, Node<V>> lookup;
    private Map<Node<V>, K> reverseLookup;

    private final int capacity;

    public LRU(int capacity) {
        this.capacity = capacity;
        this.length = 0;
        this.head = null;
        this.tail = null;
        this.lookup = new HashMap<>();
        this.reverseLookup = new HashMap<>();
    }

    public void update(K key, V value) {
        Node<V> node = lookup.get(key);
        if (node == null) {
            node = new Node<>();
            node.value = value;
            length++;
            prepend(node);
            trimCache();
            lookup.put(key, node);
            reverseLookup.put(node, key);
        } else {
            detach(node);
            prepend(node);
            node.value = value;
        }
    }

    public V get(K key) {
        Node<V> node = lookup.get(key);
        if (node == null) {
            return null;
        }
        detach(node);
        prepend(node);
        return node.value;
    }

    private void detach(Node<V> node) {
        if (node.prev != null) {
            node.prev.next = node.next;
        }
        if (node.next != null) {
            node.next.prev = node.prev;
        }
        if (length == 1) {
            head = tail = null;
        }
        if (head == node) {
            head = head.next;
        }
        if (tail == node) {
            tail = tail.prev;
        }

        node.next = null;
        node.prev = null;
    }

    private void prepend(Node<V> node) {
        if (head == null) {
            head = tail = node;
            return;
        }

        node.next = head;
        head.prev = node;
        head = node;
    }

    private void trimCache() {
        if (length <= capacity) {
            return;
        }
        Node<V> tailNode = tail;
        detach(tail);

        K key = reverseLookup.get(tailNode);
        lookup.remove(key);
        reverseLookup.remove(tailNode);
        length--;
    }
}
